core/src/cost_model.rs

```diff
@@ -0,0 +1,136 @@
+//! `cost_model` aiming to limit the size of broadcasting sets, and reducing the number
+//! of un-parallelizeble transactions (eg, transactions as same writable key sets).
+//! By doing so to improve leader performance.
+
+use solana_sdk::{
+    bpf_loader, bpf_loader_deprecated, bpf_loader_upgradeable, pubkey::Pubkey, system_program,
+};
+use solana_runtime::{
+    hashed_transaction::HashedTransaction,
+};
+use std::{collections::HashMap, str::FromStr};
+
+// TODO  revisit these hardcoded numbers, better get from mainnet log
+const COST_UNIT: u32 = 1;
+const DEFAULT_PROGRAM_COST: u32 = COST_UNIT * 500;
+const CHAIN_MAX_COST: u32 = COST_UNIT * 100_000;
+const BLOCK_MAX_COST: u32 = COST_UNIT * 100_000_000;
+
+use Pubkey as ProgramKey;
+use u32 as Cost;
+
+#[derive(Debug)]
+pub struct CostModel {
+    cost_metrics: HashMap<ProgramKey, Cost>,
+    chain_max_cost: Cost,
+    block_max_cost: Cost,
+}
+
+macro_rules! costmetrics {
+    ($( $key: expr => $val: expr ),*) => {{
+        let mut hashmap: HashMap< Pubkey, u32 > = HashMap::new();
+        $( hashmap.insert( $key, $val); )*
+        hashmap
+    }}
+}
+
+impl CostModel {
+    pub fn new() -> Self {
+        // NOTE: message.rs has following lazy_static program ids. Can probably use them to define
+        // `cost` for each type.
+        // NOTE: since each instruction has `compute budget`, possible to derive `cost` from `budget`?
+        let parse = |s| Pubkey::from_str(s).unwrap();
+        Self {
```

Comment:
stole the list of program IDs from `message.rs`. Is this a comprehensive list for all program IDs? for BPF programs, can they be identified (and cost assigned) by `bpf_loader` IDs? 

In below `find_instruction_cost` function, I added logic that is a instrument's program ID is not found in this metrics, it'd be assigned to default cost. Does this make sense?